plugin/evm/warp_backend.go

```diff
@@ -0,0 +1,82 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package evm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/database/versiondb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/ava-labs/avalanchego/vms/platformvm/teleporter"
+)
+
+var (
+	_ WarpBackend = &warpBackend{}
+
+	dbPrefix = []byte("warp")
+)
+
+// WarpBackend keeps track of messages that are accepted by the warp precompiles and add them into a database.
+// The backend is also used to query for warp message signatures by the signature request handler.
+type WarpBackend interface {
+	// AddMessage is called in the precompile OnAccept, to add warp messages into the database.
+	AddMessage(ctx context.Context, unsignedMessage *teleporter.UnsignedMessage) error
+
+	// GetSignature returns the signature of the requested message hash.
+	GetSignature(ctx context.Context, messageHash ids.ID) ([]byte, error)
+}
+
+// warpBackend implements WarpBackend, keeps track of warp messages, and generates message signatures.
+type warpBackend struct {
+	database.Database
+	snowCtx        *snow.Context
+	signatureCache *cache.LRU
+}
+
+// NewWarpBackend creates a new warpBackend, and initializes the signature cache and message tracking database.
+func NewWarpBackend(snowCtx *snow.Context, vmDB *versiondb.Database, signatureCacheSize int) WarpBackend {
+	return &warpBackend{
+		Database:       prefixdb.New(dbPrefix, vmDB),
+		snowCtx:        snowCtx,
+		signatureCache: &cache.LRU{Size: signatureCacheSize},
+	}
+}
+
+func (w *warpBackend) AddMessage(ctx context.Context, unsignedMessage *teleporter.UnsignedMessage) error {
+	messageHashBytes := hashing.ComputeHash256(unsignedMessage.Bytes())
```

Comment:
nit: can we use `hashing.ComputeHash256Array` since this will already return a `[32]byte` type ?

Reply:
Then we will not need to call `ids.ToID`

Reply:
changed to use `hashing.ComputeHash256Array` which returns the `Hash256` type. When putting in the cache i cast it `ids.ID(messageID`, because during testing when putting `Hash256` as key, i would not cache hit when getting a `ids.ID` type of same message bytes.